cmds/dhclient/dhclient.go

```diff
@@ -76,166 +75,57 @@ func ifup(ifname string) (netlink.Link, error) {
-			netmask = packet.YIAddr()
-			debug("No OptionSubnetMask; default to %v\n", netmask)
+			packet, err = client.Renew(packet)
 		}
-
-		dst := &netlink.Addr{IPNet: &net.IPNet{IP: packet.YIAddr(), Mask: netmask}, Label: ""}
-		// Add the address to the iface.
-		if *test == false {
-			if err := netlink.AddrReplace(iface, dst); err != nil {
-				if os.IsExist(err) {
-					return fmt.Errorf("add/replace %v to %v: %v", dst, iface, err)
-				}
-			}
-
-			if gwData, ok := o[dhcp4.OptionRouter]; ok {
-				debug("router %v", gwData)
-				routerName := net.IP(gwData).String()
-				debug("routerName %v", routerName)
-				r := &netlink.Route{
-					LinkIndex: iface.Attrs().Index,
-					Gw:        net.IP(gwData),
-				}
-
-				if err := netlink.RouteAdd(r); err != nil {
-					if os.IsExist(err) {
-						if err := netlink.RouteReplace(r); err != nil {
-							return fmt.Errorf("%s: add %s: %v", iface.Attrs().Name, r.String(), routerName)
-						}
-					} else {
-						return fmt.Errorf("%s: add %s: %v", iface.Attrs().Name, r.String(), routerName)
-					}
-				}
-			}
-			if ip, ok := o[dhcp4.OptionDomainNameServer]; ok {
-				rc := ""
-				// multiples of 4 octets.
-				for i := 0; i < len(ip); i += 4 {
-					// Don't let broken servers cause us to die.
-					if len(ip[i:]) < 4 {
-						log.Printf("dhcp4.OptionDomainNameServer: short length for last adddress: %v", ip[i:])
-						continue
-					}
-					rc = fmt.Sprintf("%snameserver %s\n", rc, net.IP(ip[i:i+4]))
-				}
-				if err := ioutil.WriteFile("/etc/resolv.conf", []byte(rc), 0644); err != nil {
-					return err
-				}
-			}
+		if err != nil {
+			return err
 		}
-		if binary.BigEndian.Uint16(packet.Secs()) == 0 {
-			debug("%v: server returned infinite lease.", iface.Attrs().Name)
-			break
+
+		if err := dhclient.Configure4(iface, packet); err != nil {
+			return err
 		}
 
-		// We can not assume the server will give us any grace time. So
-		// sleep for just a tiny bit less than the minimum.
-		time.Sleep(timeout - slop)
+		time.Sleep(time.Duration(*renewalTimeout) * time.Second)
 	}
 	return nil
 }
 
-// dhcp6 support in go is hard to find. This function represents our best current
-// guess based on reading and testing.
-func dhclient6(iface netlink.Link, numRenewals int, timeout time.Duration, retry int) error {
-	conn, err := dhcp6.NewPacketSock(iface.Attrs().Index)
+func DHClient6(iface netlink.Link, timeout time.Duration, retry int, numRenewals int) error {
+	client, err := dhcp6client.New(iface,
+		dhcp6client.WithTimeout(timeout),
+		dhcp6client.WithRetry(retry))
 	if err != nil {
-		return fmt.Errorf("client connection generation: %v", err)
+		return err
 	}
-	client := dhcp6.New(iface.Attrs().HardwareAddr, conn, timeout, retry)
 
-	for i := 0; numRenewals < 0 || i < numRenewals+1; i++ {
-		debug("Start getting or renewing DHCPv6 lease")
-		iaAddrs, packet, err := client.Solicit()
+	for i := 0; numRenewals < 0 || i <= numRenewals; i++ {
+		iana, packet, err := client.RapidSolicit()
```

Comment:
this will send a rapid-commit Solicit at every renewal attempt - shouldn't this send a RENEW message instead, after the first Solicit/Reply?

Reply:
It should, yeah. Thought this gets you the same lease again anyway, too.